File: March_2024/D4/Solution.java
package March_2024.D4;

import java.util.Arrays;

//Definition for singly-linked list.

public class Solution {
    public static void main(String[] args) {

    }

    public int bagOfTokensScore(int[] tokens, int power) {
        Arrays.sort(tokens);
        int left = 0;
        int score = 0;
        int maxScore = 0;
        int right = tokens.length - 1;
        while (right >= left) {
            if (tokens[left] <= power) {
                power = power - tokens[left];
                left = left + 1;
                score += 1;
            } else if (score >= 1) {
                power = power + tokens[right];
                right = right - 1;
                score -= 1;
            } else {
                break;
            }
            maxScore = Math.max(maxScore, score);
        }
        return maxScore;
    }
}

File: March_2024/D4/Solution.py
class Solution(object):
    def bagOfTokensScore(self, tokens, power):
        """
        :type tokens: List[int]
        :type power: int
        :rtype: int
        """
        tokens.sort()
        score = 0
        maxScore = 0
        left = 0
        right = len(tokens) - 1
        while right >= left:
            if tokens[left] <= power:
                power = power - tokens[left]
                left += 1
                score += 1
            elif score >= 1:
                power = power + tokens[right]
                right -= 1
                score -= 1
            else:
                break
            maxScore = max(maxScore, score)

        return maxScore